tests/test_browser.go

```diff
@@ -18,168 +21,159 @@ import (
-	vu       *k6test.VU
-	logCache *logCache
-
-	pid   int // the browser process ID
-	wsURL string
+	ctx    context.Context
+	cancel context.CancelFunc
+	vu     *k6test.VU
 
 	browserType *chromium.BrowserType
+	pid         int // the browser process ID
+	wsURL       string
 
-	api.Browser
+	*common.Browser
 
-	cancel context.CancelFunc
+	// isBrowserTypeInitialized is true if the browser type has been
+	// initialized with a VU. Some options can only be used in the
+	// post-init stage and require the browser type to be initialized.
+	isBrowserTypeInitialized bool
+
+	// http is set by the withHTTPServer option.
+	http *k6httpmultibin.HTTPMultiBin
+	// logCache is set by the withLogCache option.
+	logCache *logCache
+	// lookupFunc is set by the withEnvLookup option.
+	lookupFunc env.LookupFunc
+	// samples is set by the withSamples option.
+	samples chan k6metrics.SampleContainer
+	// skipClose is set by the withSkipClose option.
+	skipClose bool
 }
 
 // newTestBrowser configures and launches a new chrome browser.
 //
 // It automatically closes it when `t` returns unless `withSkipClose` option is provided.
 //
 // The following opts are available to customize the testBrowser:
-//   - withHTTPServer: enables the HTTPMultiBin server.
+//   - withEnvLookup: provides a custom lookup function for environment variables.
 //   - withFileServer: enables the HTTPMultiBin server and serves the given files.
+//   - withHTTPServer: enables the HTTPMultiBin server.
 //   - withLogCache: enables the log cache.
-//   - withSamplesListener: provides a channel to receive the browser metrics.
-//   - env.LookupFunc: provides a custom lookup function for environment variables.
+//   - withSamples: provides a channel to receive the browser metrics.
 //   - withSkipClose: skips closing the browser when the test finishes.
-func newTestBrowser(tb testing.TB, opts ...any) *testBrowser {
+func newTestBrowser(tb testing.TB, opts ...func(*testBrowser)) *testBrowser {
 	tb.Helper()
 
-	// set default options and then customize them
-	var (
-		enableHTTPMultiBin = false
-		enableFileServer   = false
-		enableLogCache     = false
-		skipClose          = false
-		samples            = make(chan k6metrics.SampleContainer, 1000)
-		// default lookup function is env.Lookup so that we can
-		// pass the environment variables while testing, i.e.: K6_BROWSER_LOG.
-		lookupFunc = env.Lookup
-	)
-	for _, opt := range opts {
-		switch opt := opt.(type) {
-		case httpServerOption:
-			enableHTTPMultiBin = true
-		case fileServerOption:
-			enableFileServer = true
-			enableHTTPMultiBin = true
-		case logCacheOption:
-			enableLogCache = true
-		case skipCloseOption:
-			skipClose = true
-		case withSamplesListener:
-			samples = opt
-		case env.LookupFunc:
-			lookupFunc = opt
-		}
-	}
-
-	vu := setupHTTPTestModuleInstance(tb, samples)
+	tbr := &testBrowser{t: tb}
+	tbr.applyDefaultOptions()
+	tbr.applyOptions(opts...) // apply pre-init stage options.
```

Comment:
We can also join these together like this: `applyDefaultOptions(opts...)` calls `applyOptions()` underneath. In this case, it'd make more sense to name `applyDefaultOptions` to `applyPreInitOptions`.